packages/react-aria-components/src/Button.tsx

```diff
@@ -77,31 +112,116 @@ export const ButtonContext = createContext<ContextValue<ButtonContextValue, HTML
     ...props,
-    values: {isHovered, isPressed, isFocused, isFocusVisible, isDisabled: props.isDisabled || false},
+    values: {isHovered, isPressed, isFocused, isFocusVisible, isDisabled: props.isDisabled || false, isPendingVisible},
     defaultClassName: 'react-aria-Button'
   });
 
+  let formatter = useLocalizedStringFormatter(intlMessages, 'react-aria-components');
+  let {isPending, pendingDelay = 1000, renderPendingState} = ctx;
+  // an aria label will block children and their labels from being read, this is undesirable for pending state
+  let hasAriaLabel = !!buttonProps['aria-label'] || !!buttonProps['aria-labelledby'];
+  let spinnerId = useId();
+  let buttonId = useId(buttonProps.id);
+  let containerId = useId();
+
+  useEffect(() => {
+    let timeout: ReturnType<typeof setTimeout>;
+
+    if (isPending) {
+      // Start timer when isPending is set to true.
+      timeout = setTimeout(() => {
+        setIsPendingVisible(true);
+      }, pendingDelay);
+    } else {
+      // Exit loading state when isPending is set to false. */
+      setIsPendingVisible(false);
+    }
+    return () => {
+      // Clean up on unmount or when user removes isPending prop before entering loading state.
+      clearTimeout(timeout);
+    };
+  }, [isPending]);
+
+  const isPendingAriaLiveLabel = `${hasAriaLabel ? buttonProps['aria-label'] : ''} ${formatter.format('pending')}`.trim();
+  const isPendingAriaLiveLabelledby = hasAriaLabel ? (buttonProps['aria-labelledby']?.replace(buttonId, spinnerId) ?? spinnerId) : `${containerId} ${spinnerId}`.trim();
+
+  let ariaLive: 'off' | 'polite' | 'assertive' = 'polite';
+  if (isAppleDevice() && (!hasAriaLabel || isFirefox())) {
+    ariaLive = 'off';
+  }
+
   return (
     <button
       {...filterDOMProps(props, {propNames: additionalButtonHTMLAttributes})}
       {...mergeProps(buttonProps, focusProps, hoverProps)}
       {...renderProps}
+      id={buttonId}
       ref={ref}
+      aria-label={isPending ? isPendingAriaLiveLabel : buttonProps['aria-label']}
+      aria-labelledby={isPending ? isPendingAriaLiveLabelledby : buttonProps['aria-labelledby']}
       slot={props.slot || undefined}
+      aria-disabled={isPending ? 'true' : undefined}
       data-disabled={props.isDisabled || undefined}
       data-pressed={ctx.isPressed || isPressed || undefined}
       data-hovered={isHovered || undefined}
       data-focused={isFocused || undefined}
-      data-focus-visible={isFocusVisible || undefined} />
+      data-focus-visible={isFocusVisible || undefined}>
+      <div
+        style={{display: 'contents', visibility: isPendingVisible ? 'hidden' : 'visible'}}
+        id={containerId}
+        aria-atomic>
+        {renderProps.children}
+      </div>
+      {isPending && (
+        <div
+          aria-hidden="true"
+          style={{visibility: isPendingVisible ? 'visible' : 'hidden'}}
+          className={'react-aria-ButtonPending'}>
+          {renderPendingState}
```

Comment:
Wondering if we could just provide an `isPending` render prop and let the user use that to swap the children or add an additional element conditionally? Can we make that work with the accessibility requirements?

Right now you wouldn't be able to style this extra div except via the `react-aria-ButtonPending` class, e.g. with tailwind or style macros it wouldn't be possible. And it also might be too opinionated to always reserve space for the contents - sometimes pending buttons might want to change size with an animation or something.

Reply:
yeah, i had that question about the reserved space as well https://github.com/adobe/react-spectrum/pull/6435#issue-2311863931